Add EIP: ERC-721 with transaction validation step. (#6997)
ethereum/EIPs: EIPS/eip-6997.md

---
eip: 6997
title: ERC-721 with transaction validation step.
description: A new validation step for transfer and approve calls, achieving a security step in case of stolen wallet.

> author: eduardfina (@eduardfina), Eduard López i Fina

Reviewer: If `eduardfina` and `Eduard López i Fina` are the same person, prefer:

```suggestion
author: Eduard López i Fina (@eduardfina)
```

discussions-to: https://ethereum-magicians.org/t/erc721-with-a-validation-step/14071
status: Draft
type: Standards Track
category: ERC
created: 2023-05-07
requires: 721
---

## Abstract

> This standard is an extension of [ERC-721](./eip-721.md). It defines a new validation functionality to avoid wallet draining, every `transfer` or `approve` will be locked waiting for validation.

Reviewer: ```suggestion
This standard is an extension of [ERC-721](./eip-721.md). It defines new validation functionality to avoid wallet draining: every `transfer` or `approve` will be locked waiting for validation.
```

## Motivation

> The power of the Blockchain is at the same time its weakness, giving the user full responsibility for their data.

Reviewer: ```suggestion
The power of the blockchain is at the same time its weakness: giving the user full responsibility for their data.
```

> Many cases of NFT theft currently exist, and current NFT anti-theft schemes, such as transferring NFTs to cold wallets, make NFTs inconvenient to be used.

Reviewer: ```suggestion
Many cases of NFT theft currently exist, and current NFT anti-theft schemes, such as transferring NFTs to cold wallets, make NFTs inconvenient to use.
```

> Having a validation step before every `transfer` and `approve` would bring the Smart Contract developers the possibility to create a security NFT anti-theft schemes.

Reviewer: ```suggestion
Having a validation step before every `transfer` and `approve` would give smart contract developers the opportunity to create secure NFT anti-theft schemes.
```

> When I was developing this standard I was thinking of a system where there would be a validator address that would be responsible for validating all Smart Contract transactions.
> 
> This address would be connected to a dApp where the user could see the validation requests of his NFTs and accept the correct ones.
> 
> Giving this address only the power to validate transactions would make a much more secure system where to steal an NFT the thief would have to have both the user's address and the validator address simultaneously.

Reviewer: ```suggestion
One possible use case is a system where a validator address is responsible for validating all smart contract transactions. This address would be connected to a dApp where the user could see the validation requests of his NFTs and accept the correct ones. Giving this address only the power to validate transactions would make a much more secure system where to steal an NFT the thief would have to have both the user's address and the validator address simultaneously.
```

## Specification

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.

[ERC-721](./eip-721.md) compliant contracts MAY implement this EIP.

> When `transfer` is called by the owner, the transfer MUST be frozen and a `TransferValidation` MUST be created.

Reviewer: What does "frozen" mean in this context? Does `transfer` revert, or return successfully?

Author: As you can see at the example implementation, the ERC-721 `_transfer` function is override. 
Instead of transferring the token it creates a validateTransfer object.
Then with the function _validateTransfer is when the ERC-721 `_transfer` function is called.
The same happen with `_approve` and `_setApprovalForAll `

This step of waiting for validation is what I call "frozen".

Reviewer: > As you can see at the example implementation

We prefer if the standard is implementable without having to refer to the reference implementation. All of the externally visible behaviour should be defined within the specification section. You could add something like:

```markdown
A "frozen" operation, like a `transfer`/`safeTransferFrom`/`approve`/etc. SHALL
return <!-- insert text here --> and SHALL NOT change the state of the contract
aside from <!-- insert text here -->.
```

This is just an example. Please write it as appropriate for your proposal, and include it in your document.

> the ERC-721 `_transfer` function is override.

Ah, okay. The Open Zeppelin `_transfer`, `_approve`, and similar functions aren't actually part of the ERC-721 specification. They are an implementation detail. In other words, some implementations of ERC-721 don't actually have those functions. For example, [solmate's ERC-721 implementation](https://github.com/transmissions11/solmate/blob/main/src/tokens/ERC721.sol) doesn't have any of them.

Author: Thank you, I changed the specification 👍

> When `approve` is called by the owner, the transfer MUST be frozen and an `ApproveValidation` MUST be created.

Reviewer: `approve` doesn't transfer anything. Do you mean:

```suggestion
When `approve` is called by the owner, the approval MUST be frozen and an `ApproveValidation` MUST be created.
```

> When `approve` is called by the owner, the transfer MUST be frozen and an `ApproveValidation` MUST be created.

Reviewer: I'd recommend `ApprovalValidation` instead of `ApproveValidation` because `ApproveValidation` implies that "a validation was approved", instead of "an approval needs to be validated".

Author: Changed 👍

> When `setApprovalForAll` is called by the owner, the transfer MUST be frozen and an `ApproveValidation` MUST be created.

Reviewer: ```suggestion
When `setApprovalForAll` is called by the owner, the approval MUST be frozen and an `ApproveValidation` MUST be created.
```

The `approveValidation(uint256 approveId)` function MAY be implemented as `public` or `external`.

The `totalTransferValidations()` function MAY be implemented as `pure` or `view`.

The `totalApproveValidations()` function MAY be implemented as `pure` or `view`.

## Rationale

### Universality

The standard only defines the validation functions, but not how they should be used.

Like the [ERC-721](./eip-721.md) standard with the `_mint` and `_burn` functions, it defines the validations as internal and lets the user decide how to manage them.

An example could be to have an address validator connected to a dApp so that users could manage their validations.

This validator could be used for all NFTs or only for some users.

It could also be used as an wrapped Smart Contract for existing ERC-721, allowing 1/1 conversion with existing NFTs.

### Extensibility

This standard only defines the validation function, but does not define the system with which it has to be validated. A third-party protocol can define how it wants to call these functions as it wishes.

## Backwards Compatibility

> This standard is fully [ERC-721](./eip-721.md) compatible.

Reviewer: I seriously doubt that this is fully backwards compatible. Please revisit this section after defining what "frozen" means above.

Author: The ERC-6997 adapt the ERC-721 to have an extra validation step, it has all the ERC-721 functions and only override `_transfer`, `_approve` and `_setApprovalForAll`.

I guess if it override something of the ERC-721 we can't consider it's fully backwards compatible? Still if it has the same functions with same signature.

I misunderstood the backwards compatibility, I can change it if you want 👍

Reviewer: You want to consider more than just the Solidity interface when you're writing the backwards compatibility section. The way you have "frozen" defined right now would likely not be backwards compatible with ERC-721 contracts: for example, if `transferFrom` returns successfully, they'd expect the token to be transferred.

Author: Changed

Having a validation step before every `transfer` and `approve` would give smart contract developers the opportunity to create secure NFT anti-theft schemes.

When I was developing this standard I was thinking of a system where there would be a validator address that would be responsible for validating all Smart Contract transactions.

This address would be connected to a dApp where the user could see the validation requests of his NFTs and accept the correct ones.

Giving this address only the power to validate transactions would make a much more secure system where to steal an NFT the thief would have to have both the user's address and the validator address simultaneously.

## Specification

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.

[ERC-721](./eip-721.md) compliant contracts MAY implement this EIP.

All the operations that change the ownership of an NFT, like a `transferFrom`/`safeTransferFrom`, SHALL create a `TransferValidation` pending to be validated and SHALL NOT transfer the ownership of an NFT.

> All the operations that enable an approval to manage an NFT, like an `approve`/`setApprovalForAll`, SHALL create an `ApprovalValidation` pending to be validated and SHALL NOT enable an approval.

Author: You are right, this functions shall emit a `ValidateTransfer` or `ValidateApproval` event, I added it in the last commits.

        v.tokenId = tokenId;

        _approvalValidations[_totalApprovalValidations] = v;

        emit ValidateApproval(v.owner, to, tokenId, false, _totalApprovalValidations);

        _totalApprovalValidations++;
    }

    /**
     * @dev If approved is true create an approval petition from `operator` to operate on
     * all of `owner` tokens, if not remove `operator` from operate on all of `owner` tokens
     *
     * Emits an {ValidateApproval} event.
     */
    function _setApprovalForAll(
        address owner,
        address operator,
        bool approved
    ) internal override virtual {
        require(owner != operator, "ERC6997: approve to caller");

        if(approved) {
            ApprovalValidation memory v;

            v.owner = owner;
            v.approve = operator;
            v.approveAll = true;

            _approvalValidations[_totalApprovalValidations] = v;

            emit ValidateApproval(v.owner, operator, 0, true, _totalApprovalValidations);

            _totalApprovalValidations++;
        }
        else {
            super._setApprovalForAll(owner, operator, approved);
        }
    }
}
```

## Security Considerations

> The `_validateTransfer` and `_validateApproval` functions are created as internal, tries to create a secure system by which they can be called.

Reviewer: Hm, you'll want to focus more on the security considerations of the proposal as a whole (and not just of the reference implementation.) Hypothetically, if you remove the Reference Implementation, the rest of the proposal should still make sense.

Author: I understand, I edited the Security Considerations. 👍

The power of the blockchain is at the same time its weakness: giving the user full responsibility for their data.

Many cases of NFT theft currently exist, and current NFT anti-theft schemes, such as transferring NFTs to cold wallets, make NFTs inconvenient to use.

Having a validation step before every `transfer` and `approve` would give Smart Contract developers the opportunity to create secure NFT anti-theft schemes.

An implementation example would be a system where a validator address is responsible for validating all Smart Contract transactions.

This address would be connected to a dApp where the user could see the validation requests of his NFTs and accept the correct ones.

Giving this address only the power to validate transactions would make a much more secure system where to steal an NFT the thief would have to have both the user's address and the validator address simultaneously.

## Specification

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.

[ERC-721](./eip-721.md) compliant contracts MAY implement this EIP.

> All the operations that change the ownership of an NFT, like a `transferFrom`/`safeTransferFrom`, SHALL create a `TransferValidation` pending to be validated and emit a `ValidateTransfer`, SHALL NOT transfer the ownership of an NFT.

Reviewer: ```suggestion
All the operations that change the ownership of an NFT, like a `transferFrom`/`safeTransferFrom`, SHALL create a `TransferValidation` pending to be validated and emit a `ValidateTransfer`, and SHALL NOT transfer the ownership of an NFT.
```

> All the operations that enable an approval to manage an NFT, like an `approve`/`setApprovalForAll`, SHALL create an `ApprovalValidation` pending to be validated and emit a `ValidateApproval`, SHALL NOT enable an approval.

Reviewer: ```suggestion
All the operations that enable an approval to manage an NFT, like an `approve`/`setApprovalForAll`, SHALL create an `ApprovalValidation` pending to be validated and emit a `ValidateApproval`, and SHALL NOT enable an approval.
```
